lib/define/MongoDB.js

```diff
@@ -0,0 +1,218 @@
+var singleton_pools = {}
+module.exports.Context = function (config) {
+    assert.equal(config.type, 'MongoDB')
+    const config_key = crypto.createHash('md5')
+        .update(JSON.stringify(config))
+        .digest('hex')
+    if (!singleton_pools[config_key]) {
+        singleton_pools[config_key] = config
+        log.info('MongoDB info created:', config)
+    }
+    const pool = singleton_pools[config_key]
+    this.getConnection = function () {
+        log.info('getConnection MongoDB')
+        var ret = new MongoConnection(pool, config.readonly)
+        return ret
+    }
+}
+
+function MongoConnection(pool, readonly) {
+    this._pool = pool.DB_CONN_STR
+    this._host = pool.host
+    this._port = pool.port
+    this._db = pool.db;
+}
+var user = require('./user')
+/**
+ * select data satified the condition
+ * @param set collection
+ * @param where condition
+ * @param return -1 if error ,otherwise return data
+ */
+MongoConnection.prototype.select = function (set, where, callback) {
+    if (this._readonly) {
+        return errNotAllowed(callback)
+    }
+    var server = new mongodb.Server(this._host, this._port, { auto_reconnect: true });
+    var db = new mongodb.Db(this._db, server, { safe: true });
+
+    db.open(function (err, db) {

```

Comment:
there should be NO duplication of source code. `server ... db.open(...)` is duplicated multiple times